File: scripts/generate-news.mjs
#!/usr/bin/env node
/**
 * Genera les notícies diàries per a InfoPol usant Claude API amb cerca web.
 * S'executa cada dia a les 22:00 via GitHub Actions.
 * Actualitza: src/data/noticias.js
 */

import Anthropic from '@anthropic-ai/sdk';
import { writeFileSync, readFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));
const ROOT = join(__dirname, '..');

const client = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY });

const today = new Date().toISOString().slice(0, 10);
const dateLabel = today.slice(5).replace('-', '·');

const PROMPT = `Ets el redactor de notícies per a InfoPol, portal d'informació per a agents de les forces de seguretat de Catalunya.

Avui és ${today}. Busca les notícies més rellevants del dia de Catalunya, Espanya i l'àmbit internacional.

Has de cobrir OBLIGATÒRIAMENT les categories següents (almenys 1-2 notícies per categoria):
- política: Catalunya (Generalitat, Parlament), Espanya (Govern central, Congrés), Internacional
- economia: mercat laboral, empreses, mercats, habitatge, inflació
- cultura: arts, literatura, música, cinema, patrimoni, societat
- esports: futbol, bàsquet, atletisme, tennis, esports catalans
- policial: detencions, operacions policials, judicis, sentències, successos greus
- descobriments: ciència, tecnologia, medicina, medi ambient
- premis: reconeixements, guardons, distincions, certàmens

Per a cada notícia genera un objecte JSON amb EXACTAMENT aquest format (sense afegir ni treure camps):
{
  "id": "n-YYYYMMDD-NNN",
  "date": "${today}",
  "dateLabel": "${dateLabel}",
  "categoria": "política|economia|cultura|esports|policial|descobriments|premis",
  "title": "Títol breu i directe (màx 80 caràcters)",
  "desc": "Resum objectiu en 2-3 frases. Explica el fet clau, el context i l'impacte. Màx 250 caràcters.",
  "url": "URL directa a la notícia original (font fiable: elpais.com, lavanguardia.com, ccma.cat, elconfidencial.com, etc.)"
}

On NNN és el número correlatiu de 3 dígits (001, 002, 003...).

Genera entre 12 i 16 notícies. Prioritza notícies d'avui ${today}. Usa fonts fiables.
Retorna ÚNICAMENT l'array JSON vàlid, sense cap text addicional, sense markdown, sense blocs de codi.`;

async function loadExisting() {
  const dataPath = join(ROOT, 'src/data/noticias.js');
  if (!existsSync(dataPath)) return [];
  const content = readFileSync(dataPath, 'utf-8');
  const match = content.match(/export const NOTICIAS = (\[[\s\S]*?\]);/);
  if (!match) return [];
  try {
    return JSON.parse(match[1]).filter(n => n.date !== today);
  } catch {
    return [];
  }
}

async function runWithRetry(fn, retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      return await fn();
    } catch (err) {
      if (i === retries - 1) throw err;
      const wait = Math.pow(2, i + 1) * 1000;
      console.warn(`Intent ${i + 1} fallat: ${err.message}. Reintentant en ${wait / 1000}s...`);
      await new Promise(r => setTimeout(r, wait));
    }
  }
}

async function generateNews() {
  console.log(`\n📰 Generant notícies per al ${today}...\n`);

  const response = await runWithRetry(() =>
    client.messages.create({
      model: 'claude-opus-4-7',
      max_tokens: 8192,
      tools: [{ type: 'web_search', name: 'web_search' }],
      messages: [{ role: 'user', content: PROMPT }],
    })
  );

  // Extract final text (after tool use turns)
  let finalText = '';
  for (const block of response.content) {
    if (block.type === 'text') finalText = block.text.trim();
  }

  if (!finalText) throw new Error('Resposta buida de la API');

  // Strip markdown code fences if present
  finalText = finalText.replace(/^```(?:json)?\s*/i, '').replace(/\s*```$/i, '').trim();

  let newsItems;
  try {
    newsItems = JSON.parse(finalText);
  } catch {
    // Try to extract JSON array from text
    const match = finalText.match(/\[[\s\S]*\]/);
    if (!match) throw new Error('No s\'ha pogut extreure JSON de la resposta');
    newsItems = JSON.parse(match[0]);
  }

  if (!Array.isArray(newsItems) || newsItems.length === 0) {
    throw new Error('La resposta no conté cap notícia vàlida');
  }

  // Merge with existing (keep last 30 days of previous news)
  const existing = await loadExisting();
  const merged = [...newsItems, ...existing].slice(0, 60);

  const dataPath = join(ROOT, 'src/data/noticias.js');
  const fileContent = `// Actualitzat automàticament el ${today} a les 22:00\nexport const NOTICIAS = ${JSON.stringify(merged, null, 2)};\n`;
  writeFileSync(dataPath, fileContent, 'utf-8');

  console.log(`✅ ${newsItems.length} notícies noves generades per al ${today}`);
  console.log(`📦 Total en arxiu: ${merged.length} notícies\n`);
  newsItems.forEach((n, i) => console.log(`  ${i + 1}. [${n.categoria}] ${n.title}`));
}

generateNews().catch(err => {
  console.error('\n❌ Error generant notícies:', err.message);
  process.exit(1);
});

File: scripts/generate_news.py
#!/usr/bin/env python3
"""
Genera les notícies del dia per InfoPol cridant Claude amb web search.
S'executa des del GitHub Action daily-news.yml.
"""
import json
import os
import re
import sys
from datetime import datetime

import anthropic

TODAY = datetime.now()
DATE_STR = TODAY.strftime('%Y-%m-%d')
DATE_LABEL = TODAY.strftime('%m') + '·' + TODAY.strftime('%d')
ID_PREFIX = 'n' + TODAY.strftime('%Y%m%d')

PROMPT = f"""Ets el sistema de notícies d'InfoPol. Avui és {DATE_STR}.

Cerca les notícies més importants d'avui (últimes 24h) de:
- **Catalunya**: política, economia, cultura, societat, successos, esports
- **Espanya**: política, economia, tribunals, successos rellevants, esports
- **Internacional**: geopolítica, premis, descobriments científics, successos importants

Fonts recomanades: ACN, La Vanguardia, El País, El Punt Avui, Nació Digital, BBC, Reuters, EFE.

Retorna exactament entre 8 i 12 notícies com un array JSON pur (sense cap text addicional, sense markdown, just el JSON).

Cada element ha de tenir:
{{
  "id": "{ID_PREFIX}_01",
  "date": "{DATE_STR}",
  "dateLabel": "{DATE_LABEL}",
  "tag": "una de: Política | Economia | Cultura | Esports | Policial | Internacional | Descobriment | Premi | Societat",
  "title": "titular en català, màxim 80 caràcters",
  "desc": "resum en 1-2 frases en català, màxim 180 caràcters",
  "url": "URL directa verificada a la notícia completa"
}}

Regles:
- Tots els textos (title, desc) en català
- Mai inventes notícies: cada URL ha de ser real i verificada
- Diversitat de categories i de geografia (Catalunya / Espanya / Internacional)
- La desc ha de ser informativa, no clickbait
- Incrementa el NN del id per cada notícia: _01, _02, _03..."""

NEWS_FILE = os.path.join(os.path.dirname(__file__), '..', 'src', 'data', 'news.js')
MAX_NEWS = 50


def extract_json(text: str) -> list:
    # Intenta parsejar directament
    try:
        return json.loads(text.strip())
    except json.JSONDecodeError:
        pass
    # Extreu el primer array JSON del text
    match = re.search(r'\[[\s\S]*\]', text)
    if match:
        return json.loads(match.group(0))
    raise ValueError("No s'ha trobat cap array JSON a la resposta")


def read_existing_news(path: str) -> list:
    try:
        with open(path, encoding='utf-8') as f:
            content = f.read()
        match = re.search(r'export const NEWS = (\[[\s\S]*?\]);', content)
        if match:
            return json.loads(match.group(1))
    except Exception:
        pass
    return []


def write_news_file(path: str, news: list) -> None:
    items = []
    for n in news:
        url_val = f"'{n['url']}'" if n.get('url') else 'null'
        desc = n['desc'].replace("'", "\\'")
        title = n['title'].replace("'", "\\'")
        items.append(
            f"  {{\n"
            f"    id: '{n['id']}',\n"
            f"    date: '{n['date']}',\n"
            f"    dateLabel: '{n['dateLabel']}',\n"
            f"    tag: '{n['tag']}',\n"
            f"    title: '{title}',\n"
            f"    desc: '{desc}',\n"
            f"    url: {url_val},\n"
            f"  }}"
        )
    content = "export const NEWS = [\n" + ",\n".join(items) + "\n];\n"
    with open(path, 'w', encoding='utf-8') as f:
        f.write(content)


def main():
    client = anthropic.Anthropic()

    print(f"Generant notícies per {DATE_STR}...")

    response = client.messages.create(
        model="claude-sonnet-5",
        max_tokens=4096,
        tools=[{"type": "web_search_20250305", "name": "web_search"}],
        messages=[{"role": "user", "content": PROMPT}],
    )

    # Extreu el text final de la resposta
    text_blocks = [b.text for b in response.content if hasattr(b, 'text')]
    if not text_blocks:
        print("ERROR: No s'ha obtingut text de la resposta", file=sys.stderr)
        sys.exit(1)

    raw = text_blocks[-1]

    try:
        new_items = extract_json(raw)
    except (ValueError, json.JSONDecodeError) as e:
        print(f"ERROR parsejant JSON: {e}\nResposta:\n{raw}", file=sys.stderr)
        sys.exit(1)

    print(f"Notícies generades: {len(new_items)}")

    existing = read_existing_news(NEWS_FILE)
    # Elimina notícies del dia per evitar duplicats si s'executa dues vegades
    existing = [n for n in existing if not n.get('id', '').startswith(ID_PREFIX)]
    combined = new_items + existing
    if len(combined) > MAX_NEWS:
        combined = combined[:MAX_NEWS]

    write_news_file(NEWS_FILE, combined)
    print(f"Fitxer actualitzat: {len(combined)} notícies totals")


if __name__ == '__main__':
    main()
